### PR TITLE
Enhance AmpContext for parity with window.context

### PR DESCRIPTION
**Pull request by alanorozco**
#8087

Splits `AmpContext` into `AmpContext` and `IntegrationAmpContext`. I didn't expose everything in `AmpContext` since I believe we need to be careful in what we support in the context binary. I'd be happy to merge both classes in case we're absolutely sure all methods need to be included in `AmpContext`. The base class for these is abstract due to the Liskov substitution principle.

Summary:

- Exposes other needed context fields.
- Removes `|| null` short-circuit assignment (since some fields can be boolean).
- Parses `IframeMessagingClient`-type messages in test fixture.
- Extracts functions for integration test so it can be ran for both experiment toggle states.
- Modifies `observeIntersection` implementation to match implementation in `integration.js`

### CODE REVIEW

File: 3p/ampcontext-integration.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+      this.updateDimensionsEnabled_ = true;
+    }
+  }
+
+  /**
+   * @param {number} width
+   * @param {number} height
+   */
+  updateDimensions(width, height) {
+    dev().assert(this.updateDimensionsEnabled_, 'Not available.');
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: should be a user error

File: 3p/ampcontext-integration.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  /** @return {boolean} */
+  isMaster_() {
+    return this.master == this.win_;
+  }
+
+  /** @param {string} type */
+  setEmbedType(type) {
+    this.embedType_ = type;
+
+    if (type === 'facebook' || type === 'twitter') {
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: did you missed "github"? in fact, is this method needed if we move the check into `updateDimensions`?

File: 3p/ampcontext-integration.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  /**
+   * @param {number} width
+   * @param {number} height
+   */
+  updateDimensions(width, height) {
+    dev().assert(this.updateDimensionsEnabled_, 'Not available.');
+    this.requestResize(width, height);
+  }
+
+  notifyBootstrapLoaded() {
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: to be consistent, you can call this method `bootsrapLoaded`, or you change the following method `renderStart` to `notifyRenderStart`

File: 3p/ampcontext-integration.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  /**
+   * Reports the "entity" that was rendered to this frame to the parent for
+   * reporting purposes.
+   * The entityId MUST NOT contain user data or personal identifiable
+   * information. One example for an acceptable data item would be the
+   * creative id of an ad, while the user's location would not be
+   * acceptable.
+   * @param {string} entityId See comment above for content.
+   */
+  // DUPLICATE IN 3p/integration
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: can you move this line of comment into the JSDoc, and make it a TODO?

File: 3p/ampcontext-integration.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+   *     you use. Must be the same for all callers from all frames that want
+   *     the same result.
+   * @param {function(function(*))} work Function implementing the work that
+   *     is to be done. Receives a second function that should be called with
+   *     the result when the work is done.
+   * @param {function(*)} cb Callback function that is called when the work is
+   *     done. The first argument is the result.
+   */
+  computeInMasterFrame(global, taskId, work, cb) {
+    computeInMasterFrame(global, taskId, work, cb);
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: are we able to move the impl here? can be a separate PR.

File: 3p/ampcontext.js [hunk 6]
```diff
@@ -62,24 +110,38 @@ export class AmpContext {
    *    every time we receive a page visibility message.
    */
   observePageVisibility(callback) {
-    return this.client_.makeRequest(
-        MessageType.SEND_EMBED_STATE,
-        MessageType.EMBED_STATE,
-        callback);
+    const visibilityCallbacksLength = this.visibilityCallbacks_.length;
```
Comment (lannka) [inline 3p/ampcontext.js]: callback might not be executed if you `makeRequest` early.

2 suggested changes:
- invoke callback immediately if there was a response.
- do `makeRequest` lazily

File: 3p/ampcontext.js [hunk 7]
```diff
@@ -62,24 +110,38 @@ export class AmpContext {
...
         MessageType.INTERSECTION,
-        callback);
+        intersection => {
+          callback(intersection.changes);
+        });
+
+    // Call the callback with the value that was transmitted when the
+    // iframe was drawn. Called in nextTick, so that callers don't
+    // have to specially handle the sync case.
+    nextTick(this.win_, () => {
```
Comment (lannka) [inline 3p/ampcontext.js]: I think there is no need to run this (though we were). Client should call `window.context.initialIntersection` to get this.

@zhouyx did we update the doc a while ago? I can't find your update.

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 8]
```diff
@@ -416,7 +416,10 @@ describe('amp-a4a', () => {
         });
       });
 
-      it('should be able to create AmpContext', () => {
+      // Skipping since the window object for the iframe context cannot be
+      // accesed here. "window_" is being set to an HTMLIframeElement, not a
+      // Window object.
+      it.skip('should be able to create AmpContext', () => {
```
Comment (lannka) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: let's just remove. i don't see this test that useful

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js [hunk 9]
```diff
@@ -444,7 +447,10 @@ describe('amp-a4a', () => {
...
```
Comment (lannka) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: this as well

File: 3p/ampcontext.js (hunk 7 above)
Comment (zhouyx) [inline 3p/ampcontext.js]: We didn't update the doc, it was there from beginning [https://github.com/ampproject/amphtml/blob/master/ads/README.md#initial-position](url).

Comment (zhouyx) [inline 3p/ampcontext.js]: doc update in this PR.  https://github.com/ampproject/amphtml/pull/6374/files#diff-8e253a89bfca066fd8ca624c28468261L81

Comment (lannka) [inline 3p/ampcontext.js]: let's leave the logic here for now. it will be a breaking change to those who rely on. I'll create a deprecation announcement first.

Comment (lannka) [inline 3p/ampcontext.js]: created an issue for it #8562

File: 3p/ampcontext-integration.js (hunk 2 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Missed `github`! Thanks!

I made the method to encapsulate the special case behavior. Not exactly needed, but I think it makes things more readable in the future.

File: 3p/ampcontext-integration.js (hunk 1 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Done.

File: 3p/ampcontext-integration.js (hunk 4 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Done.

File: 3p/ampcontext-integration.js (hunk 5 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: It's being used in other modules as well. I'd be happy to test for race conditions caused by making all use `window.context.computeInMasterFrame` separately.

File: 3p/ampcontext.js (hunk 6 above)
Comment (alanorozco) [inline 3p/ampcontext.js]: True.

Moved the default behavior into the module to avoid the race condition.

File: 3p/ampcontext.js (hunk 7 above)
Comment (alanorozco) [inline 3p/ampcontext.js]: Thanks! Added a TODO to deprecate.

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 8 above)
Comment (alanorozco) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: done

File: extensions/amp-a4a/0.1/test/test-amp-a4a.js (hunk 9 above)
Comment (alanorozco) [inline extensions/amp-a4a/0.1/test/test-amp-a4a.js]: and done!

File: 3p/ampcontext-integration.js (hunk 3 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Changed to `bootsrapLoaded`.

File: 3p/ampcontext.js [hunk 10]
```diff
@@ -40,18 +52,63 @@ export class AmpContext {
...
+    this.mode = null;
+
+    /** @type {?Object} */
+    this.initialIntersection = null;
+
+    /** @type {?string} */
+    this.sourceUrl = null;
+
+    /** @type {!Array<function(Object)>} */
+    this.visibilityCallbacks_ = [];
```
Comment (lannka) [inline 3p/ampcontext.js]: you might use `Observable` from `src/observable.js`

File: 3p/ampcontext.js (hunk 6 above)
Comment (lannka) [inline 3p/ampcontext.js]: what was the new change? I don't see it

Comment (alanorozco) [inline 3p/ampcontext.js]: - Removed handler from `3p/integration.js`
- Added handler (`dispatchVisibilityChangeEvent_`) in `AmpContext.js`

Comment (lannka) [inline 3p/ampcontext.js]: I think it does not address my concern. 
we expect `callback` to be executed at least once for the current page status.

File: 3p/ampcontext.js (hunk 10 above)
Comment (alanorozco) [inline 3p/ampcontext.js]: Cool! Didn't know that existed.

File: 3p/ampcontext-integration.js [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    return this.isMaster_();
+  }
+
+  /** @return {boolean} */
+  isMaster_() {
+    return this.master == this.win_;
+  }
+
+  /** @param {string} type */
+  setEmbedType(type) {
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: we can pass in type in constructor.

File: 3p/ampcontext-integration.js [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  /** @return {boolean} */
+  get isMaster() {
+    return this.isMaster_();
+  }
+
+  /** @return {boolean} */
+  isMaster_() {
+    return this.master == this.win_;
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: same here

File: 3p/ampcontext-integration.js [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  /** @return {!Window} */
+  get master() {
+    return this.master_();
+  }
+
+  /** @return {!Window} */
+  master_() {
+    return masterSelection(this.win_, dev().assertString(this.embedType_));
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: can this method be inlined?

File: 3p/ampcontext.js [hunk 14]
```diff
@@ -40,18 +53,63 @@ export class AmpContext {
...
+    this.visibilityObservable_ = new Observable();
+
     this.findAndSetMetadata_();
+
+    /** @protected {!IframeMessagingClient} */
     this.client_ = new IframeMessagingClient(win);
     this.client_.setHostWindow(this.getHostWindow_());
     this.client_.setSentinel(dev().assertString(this.sentinel));
+
+    this.listenForPageVisibility_();
```
Comment (lannka) [inline 3p/ampcontext.js]: can we listen lazily? move this into observePageVisibility

File: 3p/ampcontext.js [hunk 15]
```diff
@@ -62,24 +120,33 @@ export class AmpContext {
...
-   *  @param {function(Object)} callback Function to call every time we receive
-   *  an intersection message.
+   *  @param {function(Array<Object>)} callback Function to call every time we
+   *    receive an intersection message.
    *  @returns {function()} that when called stops triggering the callback
    *    every time we receive an intersection message.
    */
   observeIntersection(callback) {
-    return this.client_.makeRequest(
+    const unlisten = this.client_.makeRequest(
```
Comment (lannka) [inline 3p/ampcontext.js]: should switch to observable as well, so that observeIntersection can be called multiple times.

File: 3p/ampcontext.js (hunk 14 above)
Comment (alanorozco) [inline 3p/ampcontext.js]: See in `integration.js` how the visibility listener is created immediately. (`updateVisibilityState`)

Not sure how much integration would depend on that behavior.

File: 3p/ampcontext-integration.js (hunk 13 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Nope, getters in ES6 aren't allowed to access protected properties. Not sure if it's a Closure bug or WAI.

File: 3p/ampcontext-integration.js (hunk 12 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Ditto.

File: 3p/ampcontext-integration.js (hunk 11 above)
Comment (alanorozco) [inline 3p/ampcontext-integration.js]: Then the signatures for `AmpContext` and `IntegrationAmpContext` would be different :(

Can we rely on context attributes setting the embed type? In that case we could call it from `findAndSetMetadata`

File: 3p/ampcontext.js (hunk 15 above)
Comment (alanorozco) [inline 3p/ampcontext.js]: This is odd, why does the implementation for `IframeMessagingClient` override callbacks?

Added an observable for now, but do you think it would make more sense to either do:

A. Modify `IframeMessagingClient.makeRequest` so that the requests don't override
B. Add a method in `IframeMessagingClient` that would *add* callbacks instead of overriding them?

Any of the options will avoid littering `AmpContext` with observables.

Comment (lannka) [inline 3p/ampcontext.js]: Agree. Let's move the observable logic into IframeMessagingClient.makeRequest

File: 3p/ampcontext.js (hunk 14 above)
Comment (lannka) [inline 3p/ampcontext.js]: Discussed offline, we will eventually deparate `amp:visibilitychange` event and replace it with observePageVisibility API.

We will keep this in constructor for now, and start a deprecation process once the new API is available.

File: 3p/ampcontext.js [hunk 16]
```diff
@@ -62,24 +120,33 @@ export class AmpContext {
    *    every time we receive a page visibility message.
    */
   observePageVisibility(callback) {
```
Comment (lannka) [inline 3p/ampcontext.js]: discussed offline, we'll rename it to `onPageVisibilityChange(callback)` which documents the current behavior clearer (only fires on change).

File: 3p/ampcontext.js (hunk 14 above)
Comment (alanorozco) [inline 3p/ampcontext.js]: Done

File: 3p/ampcontext-integration.js (hunk 11 above)
Comment (lannka) [inline 3p/ampcontext-integration.js]: moving to findAndSetMetadata sounds good.

File: 3p/ampcontext-integration.js [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+ * Returns the "master frame" for all widgets of a given type.
+ * This frame should be used to e.g. fetch scripts that can
+ * be reused across frames.
+ * once experiment is removed.
+ * @param {!Window} win
+ * @param {string} type
+ * @return {!Window}
+ */
+// DUPLICATE IN 3p/integration
+function masterSelection(win, type) {
```
Comment (lannka) [inline 3p/ampcontext-integration.js]: shall we export and let integration.js import from here?
